### PR TITLE
In Batch Copy, add option to disable overwriting newer target table/p…

### PR DESCRIPTION
**Pull request by zshao**
…artitions.

### CODE REVIEW

**Comment by zshao** [comment]
Note that the patch changes 2 places:

1. TaskEstimator:  This is in Batch Stage 1, when we are estimating the work to do.  We will skip the case when the src Table/Partition is older than the dest Table/Partition.
2. Copy*Task:  This is in Batch Stage 3, we do another check to make sure we don't overwrite newer table/partition right before we commit the changes.

File: main/src/main/java/com/airbnb/reair/incremental/ReplicationUtils.java [hunk 1]
```diff
@@ -497,4 +497,72 @@ public static Partition normalizeNames(Partition partition) {
...
+      return Integer.valueOf(value);
+    } catch (NumberFormatException e) {
+      LOG.warn("NumberFormatException: value for key " + key + ": " + value
+          + " is not a valid Integer.");
+      return null;
+    }
+  }
+
+  private static Integer getLastModifiedTime(Map<String, String> parameters) {
+    Integer lastModifiedTime = getIntValue(parameters, HiveParameterKeys.TLMT);
```
Comment (plypaul) [inline main/src/main/java/com/airbnb/reair/incremental/ReplicationUtils.java]: These should be `Long`s

File: main/src/main/java/com/airbnb/reair/incremental/primitives/CopyUnpartitionedTableTask.java [hunk 2]
```diff
@@ -102,6 +103,15 @@ public RunInfo runTask() throws HiveMetastoreException, DistCpException, IOExcep
 
     if (existingTable != null) {
       LOG.debug("Table " + spec + " exists on destination");
+
+      if (!conf.getBoolean(ConfigurationKeys.BATCH_JOB_OVERWRITE_NEWER, true)) {
+        Table freshDestTable = existingTable;
+        if (ReplicationUtils.isSrcOlder(freshSrcTable, freshDestTable)) {
+          LOG.warn("Source table " + spec + " is older than destination, so not copying");
```
Comment (plypaul) [inline main/src/main/java/com/airbnb/reair/incremental/primitives/CopyUnpartitionedTableTask.java]: Can you print out the modified times as well?

File: main/src/main/java/com/airbnb/reair/incremental/primitives/TaskEstimator.java [hunk 3]
```diff
@@ -95,6 +96,16 @@ private TaskEstimate analyzeTableSpec(HiveObjectSpec spec)
           Optional.empty());
     }
 
+    // If both src and dest exist, and the dest is newer, and we don't overwrite newer partitions,
+    // then it's a NO_OP.
+    if (!conf.getBoolean(ConfigurationKeys.BATCH_JOB_OVERWRITE_NEWER, true)) {
+      if (ReplicationUtils.isSrcOlder(tableOnSrc, tableOnDest)) {
+        LOG.warn("Source table " + spec + " is older than destination, so not copying");
```
Comment (plypaul) [inline main/src/main/java/com/airbnb/reair/incremental/primitives/TaskEstimator.java]: Same here

File: main/src/main/java/com/airbnb/reair/incremental/primitives/TaskEstimator.java [hunk 4]
```diff
@@ -156,6 +167,16 @@ private TaskEstimate analyzePartitionSpec(HiveObjectSpec spec)
           Optional.empty());
     }
 
+    // If both src and dest exist, and the dest is newer, and we don't overwrite newer partitions,
+    // then it's a NO_OP.
+    if (!conf.getBoolean(ConfigurationKeys.BATCH_JOB_OVERWRITE_NEWER, true)) {
+      if (ReplicationUtils.isSrcOlder(partitionOnSrc, partitionOnDest)) {
+        LOG.warn("Source partition " + spec + " is older than destination, so not copying");
```
Comment (plypaul) [inline main/src/main/java/com/airbnb/reair/incremental/primitives/TaskEstimator.java]: Same here

**Comment by plypaul** [comment]
For reference, here's an example of a test for batch replication:

https://github.com/airbnb/reair/blob/master/main/src/test/java/test/BatchReplicationTest.java#L45

**Comment by zshao** [comment]
Thanks!  I will add a unit test.

**Comment by zshao** [comment]
@plypaul  I added a test and also addressed comments on the warning message.  Can you take a look again?

File: main/src/main/java/com/airbnb/reair/incremental/primitives/CopyUnpartitionedTableTask.java [hunk 5]
```diff
@@ -102,6 +103,19 @@ public RunInfo runTask() throws HiveMetastoreException, DistCpException, IOExcep
 
     if (existingTable != null) {
       LOG.debug("Table " + spec + " exists on destination");
+
+      if (!conf.getBoolean(ConfigurationKeys.BATCH_JOB_OVERWRITE_NEWER, true)) {
+        Table freshDestTable = existingTable;
+        if (ReplicationUtils.isSrcOlder(freshSrcTable, freshDestTable)) {
+          LOG.warn("Source table " + spec + " ("
```
Comment (plypaul) [inline main/src/main/java/com/airbnb/reair/incremental/primitives/CopyUnpartitionedTableTask.java]: Great, this log line will be more useful for debugging. Could you do the same for the two other places below?

**Comment by plypaul** [comment]
Thanks @zshao ! LGTM and will merge after the minor nits are addressed.

**Comment by zshao** [comment]
Thanks Paul!

**Pull request merged**